src/coordination/coordinator_state_machine.cpp

```diff
@@ -12,14 +12,136 @@
+  if (!snapshot_ctx_str.has_value()) {
+    throw std::runtime_error("Failed to retrieve snapshot context from disk");
+  }
+
+  auto snapshot_ctx_json = nlohmann::json::parse(snapshot_ctx_str.value());
+
+  auto cluster_state_json = snapshot_ctx_json.at(kCoordClusterState.data()).get<std::string>();
+  auto last_log_idx = snapshot_ctx_json.at(kLastLogIdx.data()).get<uint64_t>();
+  auto last_log_term = snapshot_ctx_json.at(kLastLogTerm.data()).get<uint64_t>();
+  auto size = snapshot_ctx_json.at(kSize.data()).get<uint64_t>();
+  auto last_config_json = snapshot_ctx_json.at(kLastConfig.data()).get<std::string>();
+  auto type = GetSnapshotType(snapshot_ctx_json.at(kType.data()).get<int>());
+
+  auto last_config = memgraph::coordination::DeserializeClusterConfig(nlohmann::json::parse(last_config_json));
+
+  auto deserialized_snapshot = cs_new<snapshot>(last_log_idx, last_log_term, last_config, size, type);
+
+  auto cluster_state =
+      memgraph::coordination::CoordinatorClusterState::DeserializeJson(nlohmann::json::parse(cluster_state_json));
+  return cs_new<memgraph::coordination::SnapshotCtx>(deserialized_snapshot, cluster_state);
+}
+
+auto SerializeSnapshotCtxToJson(memgraph::coordination::SnapshotCtx const &snapshot_ctx) -> nlohmann::json {
+  return nlohmann::json{
+      {kCoordClusterState, snapshot_ctx.cluster_state_.SerializeJson().dump()},
+      {kLastLogTerm, snapshot_ctx.snapshot_->get_last_log_term()},
+      {kLastLogIdx, snapshot_ctx.snapshot_->get_last_log_idx()},
+      {kSize, snapshot_ctx.snapshot_->size()},
+      {kLastConfig, memgraph::coordination::SerializeClusterConfig(*snapshot_ctx.snapshot_->get_last_config()).dump()},
+      {kType, FromSnapshotType(snapshot_ctx.snapshot_->get_type())}};
+}
+
 }  // namespace
 
 namespace memgraph::coordination {
 
-CoordinatorStateMachine::CoordinatorStateMachine(LoggerWrapper logger) : logger_(logger) {}
+CoordinatorStateMachine::CoordinatorStateMachine(std::shared_ptr<kvstore::KVStore> durability, LoggerWrapper logger)
+    : logger_(logger), durability_store_(std::move(durability)) {
+  if (!durability_store_) {
+    logger_.Log(nuraft_log_level::INFO, "Coordinator state machine stores snapshots only in memory from now on.");
+    return;
+  }
+
+  logger_.Log(nuraft_log_level::INFO, "Restoring coordinator state machine with durability.");
+
+  int version{0};
+  auto maybe_version = durability_store_->Get(kSnapshotVersion);
+  if (maybe_version.has_value()) {
+    version = std::stoi(maybe_version.value());
+  } else {
+    logger_.Log(
+        nuraft_log_level::TRACE,
+        fmt::format(
+            "Assuming first start of coordinator state machine using durability as version is missing, storing current "
+            "active version {}.",
+            kActiveVersion));
+    MG_ASSERT(durability_store_->Put(kSnapshotVersion, std::to_string(kActiveVersion)),
+              "Failed to store version to disk");
+    version = 1;
+  }
```

Comment:
1. Use IILE + do we need this separated again from state_manager in a separate RocksDB instance?
2. Try to reduce duplication from state_manager? 
3. Do we need to manage separate versions for state_machine and state_manager?

Reply:
Versioning can be independent between state manager and state machine, so I think we are good there

Reply:
change to single version for logs and snapshots